### PR TITLE
Spectrogram reform

### PR DESCRIPTION
**Pull request by ATriantafyllopoulos**
### Summary

Tidy up and extend our existing Spectrogram transforms.

### Proposed Changes

* Add `n_fft` parameter to `stft` functional and `Spectrogram` transform to allow the user to optionally use a different window size than FFT size

* Add `Log` transforms that allows for `magnitude_boost` as well

* Add `standardize` functional and `Standardize` transform that take care of mean and std normalizing a signal, either globally or across an axis.

* Schedule `LogSpectrogram` transform for deprecation as it does not add any extra functionality.
### Discussion

Only if needed:

1. Should I remove `LogSpectrogram` now or should we do it in a later version
2. Should I allow the `Standardize` transform to only do mean-normalization or std-normalization based on some user options? e.g. `mean=True`, `std=True`?

### CODE REVIEW

File: tests/test_transforms.py [hunk 1]
```diff
@@ -207,23 +220,22 @@ def test_resample(input, input_sample_rate, output_sample_rate, method, axis):
...
     (np.random.normal(size=[2, 16000, 3]), 2048, 1024, 1, 1e-07),
     (np.random.normal(size=[16000, 2, 3]), 2048, 1024, 0, 1e-07),
+    (np.random.normal(size=[16000, 2, 3]), 2048, 1024, 0, 0),
     (np.random.normal(size=16000), 2048, 1024, 0, 1e-07),
 ])
 def test_stft(input, window_size, hop_size, axis, magnitude_boost):
     t = transforms.Spectrogram(window_size, hop_size, axis=axis)
-    t_log = transforms.LogSpectrogram(window_size, hop_size, axis=axis,
-                                      magnitude_boost=magnitude_boost)
+    t_log = transforms.Log(magnitude_boost=magnitude_boost)
```
Comment (hagenw) [inline tests/test_transforms.py]: I would prefer to create an extra `test_log` in order to easier find the test.

File: tests/test_transforms.py [hunk 2]
```diff
@@ -207,23 +220,22 @@ def test_resample(input, input_sample_rate, output_sample_rate, method, axis):
 @pytest.mark.parametrize('input,window_size,hop_size,axis,magnitude_boost', [
     (A, 4, 1, 2, 1e-07),
     pytest.param(A, 4, 1, 2, -1e-07, marks=xfail(raises=ValueError)),
-    pytest.param(A, 4, 1, 2, 0, marks=xfail(raises=ValueError)),
+    pytest.param(A, 4, 1, 2, -1, marks=xfail(raises=ValueError)),
```
Comment (hagenw) [inline tests/test_transforms.py]: Why did you change the axis to `-1`?

File: tests/test_transforms.py [hunk 3]
```diff
@@ -140,6 +140,19 @@ def test_normalize(input, axis, expected_output):
...
+    (A, None),
+    (A, -1),
+    (a11, None),
+    (a11, -1),
+])
+def test_standardize(input, axis):
+    t = transforms.Standardize(axis=axis)
+    output = t(input)
+    np.testing.assert_almost_equal(output.mean(axis=axis).mean(), 0)
+    np.testing.assert_almost_equal(output.std(axis=axis).mean(), 1)
```
Comment (hagenw) [inline tests/test_transforms.py]: Please also add a test for `F.standardize` in `tests/test_transforms_functional.py`.

File: audtorch/transforms/transforms.py [hunk 4]
```diff
@@ -812,6 +855,45 @@ def __repr__(self):
...
+class Log(object):
+    r"""Logarithmic transform of an input signal.
+
+    * :attr:`magnitude_boost` controls the non-negative value added to the
+      magnitude of the signal before applying the logarithmus
+
+    Args:
+        magnitude_boost (float, optional): positive value added to the
+            magnitude of the signal before applying the logarithmus.
+            Default: `1e-7`
```
Comment (hagenw) [inline audtorch/transforms/transforms.py]: Wrong indentation.

File: audtorch/transforms/transforms.py [hunk 5]
```diff
@@ -812,6 +855,45 @@ def __repr__(self):
...
+
```
Comment (hagenw) [inline audtorch/transforms/transforms.py]: the `Shape:` part is missing.

File: audtorch/transforms/transforms.py [hunk 6]
```diff
@@ -790,17 +830,20 @@ class Spectrogram(object):
 
     """
 
-    def __init__(self, window_size, hop_size, *, window='hann', axis=-1):
+    def __init__(self, window_size, hop_size, *, n_fft=None,
```
Comment (hagenw) [inline audtorch/transforms/transforms.py]: As we use `window_size` and `hop_size` at the moment, a better name might be `fft_size`?

Comment (ATriantafyllopoulos) [reply audtorch/transforms/transforms.py]: Agreed.

File: audtorch/transforms/transforms.py (hunk 5 above)
Comment (ATriantafyllopoulos) [reply audtorch/transforms/transforms.py]: It does not modify the shape. I will add it nonetheless for consistency.

File: tests/test_transforms.py (hunk 2 above)
Comment (ATriantafyllopoulos) [reply tests/test_transforms.py]: It's not the axis but the `magnitude_boost`. I knew I had forgotten something in the discussion :)

I now allow the `magnitude_boost` to be 0 just in case the user does not want to add any limit to it.

File: tests/test_transforms.py (hunk 1 above)
Comment (ATriantafyllopoulos) [reply tests/test_transforms.py]: Will do!

**Comment by ATriantafyllopoulos** [comment]
>     1. I would leave it like you did now. First you deprecate stuff for the next release and then in the release thereafter it is removed.
:+1: 
>     2. do you have a use case in mind for adding those options?
Yes, this paper: https://www.robots.ox.ac.uk/~vgg/publications/2017/Nagrani17/nagrani17.pdf

where in order to replicate the spectrogram shape with the window size they mention I had to use different `n_fft` and `window_size` parameters.
 
>     3. In addition, I was wondering if we shouldn't extent the `stft` tests to the new `n_fft` argument as well?

I was thinking the same, but we'd probably need a more comprehensive benchmark of FFT algorithms for it to make any sense. I don't see much reason to do it just for bugs.

Also, for the sake of discussion, I have forgotten to add the new classes to the docs and will do so in a follow-up commit.

**Comment by hagenw** [comment]
Regarding the discussion point 2. Thanks for the link for the a usage of the `fft_size` option. But I was more referring to your mentioned "only do mean-normalization or std-normalization" options to `Standardize`

**Comment by ATriantafyllopoulos** [comment]
> Regarding the discussion point 2. Thanks for the link for the a usage of the `fft_size` option. But I was more referring to your mentioned "only do mean-normalization or std-normalization" options to `Standardize`

Again, if you look at the results discussion in that paper, they mention results with and without std normalization (only using mean centering)

File: tests/test_transforms.py (hunk 2 above)
Comment (hagenw) [reply tests/test_transforms.py]: OK, makes sense.

**Comment by hagenw** [comment]
Yes, you are right, now I so the different normalization schemes in that paper.
i would propose to go with your suggested options of `mean=True` and `std=True`.

File: docs/api-transforms.rst [hunk 7]
```diff
@@ -95,6 +95,12 @@ Spectrogram
 .. autoclass:: Spectrogram
     :members:
 
+Log
+---
+
+.. autoclass:: Log
+    :members:
+
```
Comment (hagenw) [inline docs/api-transforms.rst]: It looks to me that `Standardize` is missing here and `standardize` in the functional part of the documentation.

Comment (ATriantafyllopoulos) [reply docs/api-transforms.rst]: Yes, I uploaded it in the latest commits.

File: audtorch/transforms/transforms.py (hunk 4 above)
Comment (ATriantafyllopoulos) [reply audtorch/transforms/transforms.py]: Is this fixed? Tests are passing for me...

File: audtorch/transforms/functional.py [hunk 8]
```diff
@@ -256,7 +256,7 @@ def normalize(signal, *, axis=None):
     return signal / np.maximum(peak, 1e-7)
 
 
-def standardize(signal, axis=None):
+def standardize(signal, axis=None, mean=True, std=True):
```
Comment (hagenw) [inline audtorch/transforms/functional.py]: So far, we had `axis` always as the last argument. It doesn't matter, but it makes reading the source code easier as you can directly spot which functions don't provide that option. Also the `*` is missing. So, please change to:
```python
def standardize(signal, *, mean=True, std=True, axis=None):
```

File: audtorch/transforms/functional.py [hunk 9]
```diff
@@ -268,6 +268,9 @@ def standardize(signal, axis=None):
         signal (numpy.ndarray): audio signal
         axis (int, optional): standardize only along the given axis.
             Default: `None`
+        mean (bool, optional): apply mean centering. Default: `True`
+        std (bool, optional): normalize by standard deviation. Default: `True`
+
```
Comment (hagenw) [inline audtorch/transforms/functional.py]: Please move above `axis` and use only one blank line before `Returns`.

File: audtorch/transforms/functional.py [hunk 10]
```diff
@@ -285,7 +288,11 @@ def standardize(signal, axis=None):
     else:
         mean = np.mean(signal)
         std = np.std(signal)
-    return (signal - mean) / np.maximum(std, 1e-7)
+    if mean:
+        signal = signal - mean
+    if std:
+        signal = signal / np.maximum(std, 1e-7)
```
Comment (hagenw) [inline audtorch/transforms/functional.py]: I think we should only do the computation if we really have to. Fortunatly, numpy `mean` and `std` seem to allow both for `axis=None` as well. So you could change it to:
```python
if mean:
    mean_signal = np.expand_dims(np.mean(signal, axis=axis), axis=axis)
    signal = signal - mean_signal
if std:
    std_signal = np.expand_dims(np.std(signal, axis=axis), axis=axis)
    signal = signal / np.maximum(std_signal, 1e-7)
```

Note: your current implementation has a bug, as you store your calculations in the variables named `mean` and `std` which are identical with the new argument names.

Comment (hagenw) [reply audtorch/transforms/functional.py]: OK, I overlooked that you renamed the intermediate variables in a later commit. So there shouldn't be an issue.
But it should be possible to not check for the `axis` and only execute the calculation if really needed.

File: audtorch/transforms/transforms.py (hunk 4 above)
Comment (hagenw) [reply audtorch/transforms/transforms.py]: Sorry, I think this was my fault.

File: audtorch/transforms/functional.py [hunk 11]
```diff
@@ -256,10 +256,50 @@ def normalize(signal, *, axis=None):
     return signal / np.maximum(peak, 1e-7)
 
 
-def stft(signal, window_size, hop_size, *, window='hann', axis=-1):
-    r"""Short-term Fourier transform.
+def standardize(signal, axis=None, mean=True, std=True):
```
Comment (phtephanx) [inline audtorch/transforms/functional.py]: I can see the need for flexibility. Being a bit picky but should not the naming `standardize` by default entail `mean=True` and `std=True`?

Comment (phtephanx) [reply audtorch/transforms/functional.py]: I think, the compactness of this implementation outweighs the fact that it's slightly counter-intuitive.

Comment (ATriantafyllopoulos) [reply audtorch/transforms/functional.py]: It is the standard behavior people are used to by `sklearn`'s `StandardScaler`. Plus standardization usually means `mean & std` normalization. Just doing mean normalization would probably be referred to as `mean centering` for example. So I see no need to change the name or demand that the user always specifies the mean and std arguments.

File: audtorch/transforms/functional.py (hunk 10 above)
Comment (hagenw) [reply audtorch/transforms/functional.py]: I saw your new update and realized that you cannot implement as I proposed as `np.expand_dims` will not work with `axis=None`. So I think your current approach is the best one.

File: audtorch/transforms/functional.py (hunk 8 above)
Comment (hagenw) [reply audtorch/transforms/functional.py]: Thanks for your changes, could you still please move `axis=None` to the last position.

File: audtorch/transforms/functional.py (hunk 11 above)
Comment (hagenw) [reply audtorch/transforms/functional.py]: @phtephanx if you are ok with the current implementation can you mark this as resolved please.

File: audtorch/transforms/functional.py (hunk 8 above)
Comment (ATriantafyllopoulos) [reply audtorch/transforms/functional.py]: Sorry, I forgot. I moved it in f2d9694
